### PR TITLE
Yossi device key driver

### PR DESCRIPTION
**Pull request by yossi2le**
### Description

This pull request is for code review of the DeviceKey driver phase 1.

### Pull request type
- [ ] Feature

### CODE REVIEW

File: drivers/DeviceKey.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    virtual ~DeviceKey();
+
+    /** Derive a new key based on the salt string. DeviceKeyType is an enum with values 128_bit and 256_bit
+     * @param isalt input buffer used to create the new key. Same input will generate always the same key
+     * @param isalt_size size of the data in salt
+     * @param output buffer to receive the derived key. Size must 16 or 32 bytes according to the key_type passed
+     * @param key_type Size of the required key (128 or 256)
+     * @return 0 on success, negative error code on failure
+     */
+    int device_key_derive_key(const unsigned char *isalt, size_t isalt_size, unsigned char *output, size_t ikey_type);
```
Comment (dannybenor) [inline drivers/DeviceKey.h]: DeviceKeytype enum, size_t ikey_type, key_type: Balagan!

File: drivers/DeviceKey.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    /** Set a device key into the NVStore
+     * @param value input buffer contain the key.
+     * @param isize size of the supplied key. must be 16Byte or 32Byte.
+     * @return 0 on success, negative error code on failure
+     */
+    int device_key_set_value(uint32_t *value, size_t isize);
```
Comment (dannybenor) [inline drivers/DeviceKey.h]: Change the name to device_inject_root_of_trust

File: drivers/DeviceKey.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * @return 0 on success, negative error code on failure
+     */
+    int read_key_from_nvstroe(uint32_t *output, size_t &size);
+
+    /** Set a device key into the NVStore
+     * @param value input buffer contain the key.
+     * @param isize the size of the input buffer.
+     * @return 0 on success, negative error code on failure
+     */
+    int write_key_to_nvstroe(uint32_t *value, size_t isize);
```
Comment (dannybenor) [inline drivers/DeviceKey.h]: nvstore

File: targets/targets.json [hunk 4]
```diff
@@ -614,7 +614,7 @@
         "macros": ["CPU_MK64FN1M0VMD12", "FSL_RTOS_MBED"],
         "inherits": ["Target"],
         "detect_code": ["0240"],
-        "device_has": ["ANALOGIN", "ANALOGOUT", "I2C", "I2CSLAVE", "INTERRUPTIN", "LOWPOWERTIMER", "PORTIN", "PORTINOUT", "PORTOUT", "PWMOUT", "RTC", "SERIAL", "SERIAL_FC", "SERIAL_ASYNCH", "SLEEP", "SPI", "SPI_ASYNCH", "SPISLAVE", "STDIO_MESSAGES", "STORAGE", "TRNG", "FLASH"],
+        "device_has": ["DEVKEY", "ANALOGIN", "ANALOGOUT", "I2C", "I2CSLAVE", "INTERRUPTIN", "LOWPOWERTIMER", "PORTIN", "PORTINOUT", "PORTOUT", "PWMOUT", "RTC", "SERIAL", "SERIAL_FC", "SERIAL_ASYNCH", "SLEEP", "SPI", "SPI_ASYNCH", "SPISLAVE", "STDIO_MESSAGES", "STORAGE", "TRNG", "FLASH"],
```
Comment (dannybenor) [inline targets/targets.json]: Please undo this change

File: drivers/DeviceKey.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/DeviceKey.h"
+#if !defined(MBEDTLS_CONFIG_FILE)
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: Don't think you need this include. The configuration of mbedtls can be done in different ways

File: drivers/DeviceKey.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+#if !defined(MBEDTLS_CMAC_C)
+#error [NOT_SUPPORTED] MBEDTLS_CMAC_C needs to be enabled for this driver
+#else
+
+int DeviceKey::device_key_derive_key(const unsigned char *salt, size_t isalt_size, unsigned char *output, size_t ikey_type)
+{
+    uint32_t key_buff[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    size_t actual_size = DEVICE_KEY_32BYTE;
+
+    //First try to read the key form NVStore
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: comment not in right place

File: drivers/DeviceKey.cpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    if (DEVICE_KEY_16BYTE != ikey_type && DEVICE_KEY_32BYTE != ikey_type)
+    {
+        return DEVICEKEY_INVALID_KEY_TYPE;
+    }
+
+    int ret = read_key_from_nvstroe(key_buff, actual_size);
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: please fix the function name to read_key_from_nvstore

File: drivers/DeviceKey.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    if (DEVICEKEY_SUCCESS != ret && DEVICEKEY_NOT_FOUND != ret)
+    {
+        return ret;
+    }
+
+    //If the key was not find in NVStore we will create it by using TRNG and then save it to NVStore
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: found

File: drivers/DeviceKey.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    if (DEVICEKEY_NOT_FOUND == ret) {
+       if (DEVICEKEY_SUCCESS != generate_key_by_trns(key_buff,actual_size)) {
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: what is trns?

File: drivers/DeviceKey.cpp [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+           return ret;
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: ret is not up-to-date. You missed generate_key_by_trns

File: drivers/DeviceKey.cpp [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+       }
+
+       if (DEVICEKEY_SUCCESS != device_key_set_value(key_buff,actual_size)) {
+           return ret;
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: ret is not up-to-date you missed device_key_set_value

File: drivers/DeviceKey.cpp [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+       }
+    }
+
+    ret = get_derive_key(key_buff, actual_size, salt, isalt_size, output, ikey_type);
```
Comment (dannybenor) [inline drivers/DeviceKey.cpp]: change the name to get_derived_key

**Review by dannybenor — CHANGES_REQUESTED** [review]

File: TESTS/mbed_drivers/device_key/main.cpp [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+/*
+ * Injection of a dummy key when there is not TRNG
+ */
+int inject_dummy_rot_key()
+{
+#if !defined(DEVICE_TRNG)
+    uint32_t key[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+
+    memset(key, 0, DEVICE_KEY_32BYTE);
+    memcpy((void*)key,(const void*)"1234567812345678",16);
```
Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: Why do you need these casts? Should work without them as well.

Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: Spaces between arguments.

File: TESTS/mbed_drivers/device_key/main.cpp [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+    int size=16;
```
Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: Spaces

File: TESTS/mbed_drivers/device_key/main.cpp [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+ */
+void device_inject_root_of_trust_16_byte_size_test()
+{
+    DeviceKey &devkey = DeviceKey::get_instance();
+    uint32_t rkey[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    uint16_t actual_size;
+    uint32_t key[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    NVStore &nvstore = NVStore::get_instance();
+
+    nvstore.init();
```
Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: init is redundant if reset is called right after.

File: TESTS/mbed_drivers/device_key/main.cpp [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+
+/*
+ * Test the consistency of derived 32 byte key result.
+ */
+void device_key_derived_key_consistency_32_byte_key_test()
+{
+    unsigned char output1[DEVICE_KEY_32BYTE];
+    unsigned char output2[DEVICE_KEY_32BYTE];
+    unsigned char empty_buffer[DEVICE_KEY_32BYTE];
+    unsigned char salt[] = { "The quick brown fox jumps over the lazy dog" };
```
Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: String initialization in this module is not uniform - once with braces, once without. Maybe it's better to lose the braces.

File: drivers/DeviceKey.cpp [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        if (DEVICEKEY_SUCCESS != ret) {
+            return ret;
+        }
+#else
+        return DEVICEKEY_NO_KEY_INJECTED;
+#endif
+
+    }
+
+    ret = get_derive_key(key_buff, actual_size, salt, isalt_size, output, ikey_type);
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: You forgot to rename it to get_derived_key

File: drivers/DeviceKey.cpp [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+    return DEVICEKEY_SUCCESS;
+}
+
+// Calculate CMAC functions - wrapper for mbedtls start/update and finish
+int DeviceKey::calc_cmac(const unsigned char *input, size_t isize, uint32_t *ikey_buff, int ikey_size, unsigned char *output)
+{
+    int ret;
+    mbedtls_cipher_context_t ctx;
+
+    mbedtls_cipher_type_t mbedtls_cipher_type=MBEDTLS_CIPHER_AES_128_ECB;
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: spaces

File: drivers/DeviceKey.cpp [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type=MBEDTLS_CIPHER_AES_256_ECB;
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: spaces

File: drivers/DeviceKey.cpp [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+int DeviceKey::get_derive_key(uint32_t *ikey_buff, size_t ikey_size, const unsigned char *isalt,
+                              size_t isalt_size, unsigned char *output, uint32_t ikey_type)
+{
+    int ret;
+    unsigned char * double_size_salt;
+
+    if (ikey_size == ikey_type || ikey_size > ikey_type) {
+        ret = this->calc_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: `this` is redundant. calling a member function is always on `this`.

File: drivers/DeviceKey.cpp [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+
+//This method is generating hardcoded 16 byte key for now!!!
+int DeviceKey::generate_key_by_trng(uint32_t *ikey_buff, size_t &size)
+{
+    if (size < DEVICE_KEY_16BYTE)
+    {
+        return DEVICEKEY_BUFFER_TO_SMALL;
+    }
+
+    memcpy((void*)ikey_buff,(const void*)"1234567812345678",16);
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: Same here - why the casts?
Spaces as well.

File: drivers/DeviceKey.h [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+    DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR = -7,
+    DEVICEKEY_ERR_CMAC_GENERIC_FAILURE    = -8,
+    DEVICEKEY_MISSING_MBEDTLS             = -9,
+    DEVICEKEY_BUFFER_TO_SMALL             = -10,
+    DEVICEKEY_NO_KEY_INJECTED             = -11,
+};
+
+#define KEY_BUFFER_SIZE 32
+#define NVSTORE_KEY_ROT 4
+
```
Comment (davidsaada) [inline drivers/DeviceKey.h]: Do both these defines need to be in the header file (I suspect not)? If they do, limit the scope by adding a prefix to their names.

File: drivers/DeviceKey.h [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+/** Use this singleton if you need to derive a new key from the device root of trust.
+ *
+ *
+ * @endcode
+ * @ingroup drivers
+ */
+
+class DeviceKey {
```
Comment (davidsaada) [inline drivers/DeviceKey.h]: If it's a singleton class, it should definitely be non copyable. Derive it from NonCopyable (see NVStore class for reference).

File: drivers/DeviceKey.cpp [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/DeviceKey.h"
+#include "mbedtls/config.h"
+#include "mbedtls/cmac.h"
+#include <nvstore.h>
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: Use quotes here.

File: drivers/DeviceKey.cpp [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::read_key_from_nvstore(uint32_t *output, size_t &size)
+{
+    uint16_t short_size = size;
+    NVStore &nvstore = NVStore::get_instance();
+    int nvStatus = nvstore.get(NVSTORE_KEY_ROT, short_size, output, short_size);
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: Using the same argument for buffer size and returned size parameters is a very bad idea. Please define an additional variable for it (also applies to all other get instances).

**Review by davidsaada — CHANGES_REQUESTED** [review]

File: TESTS/mbed_drivers/device_key/main.cpp [hunk 26]
```diff
@@ -294,10 +297,10 @@ utest::v1::status_t greentea_failure_handler(const Case *const source, const fai
     return STATUS_CONTINUE;
 }
 
-Case cases[] = { Case("Device Key - inject value wrong size"            , device_inject_root_of_trust_wrong_size_test              , greentea_failure_handler),
-                 Case("Device Key - inject value 16 byte size"          , device_inject_root_of_trust_16_byte_size_test            , greentea_failure_handler),
-                 Case("Device Key - inject value 32 byte size"          , device_inject_root_of_trust_32_byte_size_test            , greentea_failure_handler),
-                 Case("Device Key - inject value only once"             , device_inject_root_of_trust_only_once_test               , greentea_failure_handler),
+Case cases[] = { Case("Device Key - inject value wrong size"            , device_inject_root_of_trust_wrong_size_test        , greentea_failure_handler),
```
Comment (davidsaada) [inline TESTS/mbed_drivers/device_key/main.cpp]: Brace should be in a separate line

File: drivers/DeviceKey.cpp [hunk 27]
```diff
@@ -32,27 +38,25 @@ DeviceKey::~DeviceKey()
...
-#else
-
 int DeviceKey::device_key_derived_key(const unsigned char *salt, size_t isalt_size, unsigned char *output, uint16_t ikey_type)
 {
     uint32_t key_buff[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
     size_t actual_size = DEVICE_KEY_32BYTE;
 
-    if (DEVICE_KEY_16BYTE != ikey_type && DEVICE_KEY_32BYTE != ikey_type)
-    {
+    if (DEVICE_KEY_16BYTE != ikey_type && DEVICE_KEY_32BYTE != ikey_type){
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: Space before the brace

File: drivers/DeviceKey.cpp [hunk 28]
```diff
@@ -32,27 +38,25 @@ DeviceKey::~DeviceKey()
...
 
     //First try to read the key form NVStore
     int ret = read_key_from_nvstore(key_buff, actual_size);
-    if (DEVICEKEY_SUCCESS != ret && DEVICEKEY_NOT_FOUND != ret)
-    {
+    if (DEVICEKEY_SUCCESS != ret && DEVICEKEY_NOT_FOUND != ret) {
         return ret;
     }
 
+    if(DEVICE_KEY_16BYTE != actual_size && DEVICE_KEY_32BYTE != actual_size) {
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: Space between if and parentheses (search, there are a few of these)

**Review by davidsaada — CHANGES_REQUESTED** [review]

**Review by davidsaada — APPROVED** [review]

File: drivers/DeviceKey.cpp [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+#include "mbedtls/config.h"
+#include "mbedtls/cmac.h"
+#include "nvstore.h"
+#include "trng_api.h"
+
+#if !defined(MBEDTLS_CMAC_C)
+#error [NOT_SUPPORTED] MBEDTLS_CMAC_C needs to be enabled for this driver
+#else
+
+//This constant is temporary because NVStroe do not support keys enum yet.
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: NVStore

File: drivers/DeviceKey.cpp [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    if (DEVICE_KEY_32BYTE == ikey_type) {
+        ret = this->calc_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
+        if (DEVICEKEY_SUCCESS != ret) {
+            goto finish;
+        }
+
+        //Double the salt size cause cmac always return just 16 bytes
+        double_size_salt = new unsigned char[isalt_size * 2];
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp:199]: You don't free this in case of success.

File: drivers/DeviceKey.cpp [hunk 31]
```diff
@@ -0,0 +1,257 @@
...
+    } else if (DEVICE_KEY_16BYTE <= size && DEVICE_KEY_32BYTE > size) {
+        in_size = DEVICE_KEY_16BYTE;
+    } else {
+        in_size = DEVICE_KEY_32BYTE;
+    }
+
+    trng_init(&trng_obj);
+
+    int ret = trng_get_bytes(&trng_obj, (unsigned char *)output, in_size, &size);
+    if ( DEVICEKEY_SUCCESS != ret || in_size != size) {
```
Comment (davidsaada) [inline drivers/DeviceKey.cpp]: redundant space

File: drivers/DeviceKey.h [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+    DEVICEKEY_NOT_FOUND                   = -5,
+    DEVICEKEY_READ_FAILED                 = -6,
+    DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR = -7,
+    DEVICEKEY_ERR_CMAC_GENERIC_FAILURE    = -8,
+    DEVICEKEY_BUFFER_TO_SMALL             = -9,
+    DEVICEKEY_NO_KEY_INJECTED             = -10,
+    DEVICEKEY_INVALID_PARAM               = -11,
+    DEVICEKEY_TRNG_ERROR                  = -12,
+};
+
```
Comment (davidsaada) [inline drivers/DeviceKey.h:44]: Need to enclose this class with Doxygen comments. See nvstore.h for reference.

**Review by davidsaada — CHANGES_REQUESTED** [review]

**Review by davidsaada — APPROVED** [review]

**Pull request closed without merging**